posts/extend-dataframes/index.qmd

```diff
@@ -0,0 +1,283 @@
+
+Here we present both cases as well as considerations and technical details of both options. We'll demonstrate both of these cases with the subset function in R (subsetting uses a single square bracket for tabular data, `[`). First the fail-on-subsetting. Before we write the subsetting function it is useful to have a function that checks that an object of our class is valid, a so-called validator function.
+
+```{r}
+validate_birthdays <- function(x) {
+  stopifnot(
+    "input must contain 'name' and 'birthday' columns" =
+      all(c("name", "birthday") %in% colnames(x)),
+    "names must be a character" =
+      is.character(x$name),
+    "birthday must be a date" =
+      lubridate::is.Date(x$birthday)
+  )
+  invisible(x)
+}
+```
+
+
+This will return an error if the class is not valid (defined in terms of the class' invariants). 
+
+Now we can show how to error if one of the invariants are removed during subsetting. 
+
+```{r, eval=FALSE}
+`[.birthdays` <- function(x) {
+  validate_birthdays(NextMethod())
+}
+
+birthdays[, -1]
+#  Error in validate_birthdays(NextMethod()) :
+#  input must contain 'name' and 'birthday' columns
+```
+
+The second design option is the reconstruct-on-subsetting. This checks whether the class is valid, and if not downgrade the class to the superclass, in our case a data frame. This is done by not only validating the object during subsetting but to check whether it is a valid class object, and then either ensuring all of the attributes of the subclass -- in our case `<birthdays>` -- are maintained, or attributes are stripped and only the attributes of the base superclass -- in our case `data.frame` -- are kept. 
+
+
+
+::: {.callout-note}
+Important note: this section of the post relies heavily on [https://github.com/DavisVaughan/2020-06-01_dplyr-vctrs-compat](https://github.com/DavisVaughan/2020-06-01_dplyr-vctrs-compat).
+:::
+  
+  The four functions that are required to be added to ensure our class is correctly handled when invaliding it are: 
+  
+  - `birthdays_reconstruct()`
+- `birthdays_can_reconstruct()`
+- `df_reconstruct()`
+- `dplyr_reconstruct.birthdays()`
+
+We'll tackle the first three first, and then move onto to the last one as this requires some extra steps. 
+
+`birthdays_reconstruct()` is a function that contains an if-else statement to determine whether the returned object is a `<birthdays>` or `data.frame` object.
+
+```{r}
+birthdays_reconstruct <- function(x, to) {
+  if (birthdays_can_reconstruct(x)) {
+    df_reconstruct(x, to)
+  } else {
+     x <- as.data.frame(x)
+    message("Removing crucial column in `<birthdays>` returning `<data.frame>`")
+    x
+  }
+}
+```
+
+The if-else evaluation is controlled by `birthdays_can_reconstruct()`. This function determines whether after subsetting the object is a valid `<birthdays>` class. It checks whether the validator fails, in which case it returns `FALSE`, otherwise the function will return `TRUE`.
```

Comment:
This is a matter of writing style/preference, but I would first explain `birthday_can_reconstruct()` before explaining `birthday_reconstruct()`. This style follows the "define before use" formula.

Reply:
I also gave this ordering some thought and went with introducing `birthdays_reconstruct()` first as it is the less technical function compared to `birthdays_can_reconstruct()` and so gives context to `birthdays_can_reconstruct()`. I agree that it's more personal preference.